content/4_Gold/PURS.mdx

```diff
@@ -268,14 +462,217 @@ for _ in range(query_num):
 </PySection>
 </LanguageSection>
 
-### Solution - Dynamic Range Sum Queries
+## Dynamic Range Sum Queries
+
+<FocusProblem problem="rangeSum" />
+
+### Recursive Implementation
+
+**Time Complexity:** $\mathcal{O}((N + Q) \log N)$
 
 <LanguageSection>
 <CPPSection>
 
-Compared to the previous problem, all we need to change are `DEFAULT`
+Compared to the previous implementation, all we need to change are `DEFAULT`
 and how we combine the elements.
 
+```cpp
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+// BeginCodeSnip{Segment Tree}
+template <class T> class SumSegmentTree {
+  private:
+	const T DEFAULT = 0;
+
+	int len;
+	vector<T> segtree;
+
+	T combine(const T &a, const T &b) { return a + b; }
+
+	void build(const vector<T> &arr, int at, int at_left, int at_right) {
+		if (at_left == at_right) {
+			segtree[at] = arr[at_left];
+			return;
+		}
+		int mid = (at_left + at_right) / 2;
+		build(arr, 2 * at, at_left, mid);
+		build(arr, 2 * at + 1, mid + 1, at_right);
+		segtree[at] = combine(segtree[2 * at], segtree[2 * at + 1]);
+	}
+
+	void set(int ind, T val, int at, int at_left, int at_right) {
+		if (at_left == at_right) {
+			segtree[at] = val;
+			return;
+		}
+		int mid = (at_left + at_right) / 2;
+		if (ind <= mid) {
+			set(ind, val, 2 * at, at_left, mid);
+		} else {
+			set(ind, val, 2 * at + 1, mid + 1, at_right);
+		}
+		segtree[at] = combine(segtree[2 * at], segtree[2 * at + 1]);
+	}
+
+	T range_sum(int start, int end, int at, int at_left, int at_right) {
+		if (at_right < start || end < at_left) { return DEFAULT; }
+		if (start <= at_left && at_right <= end) { return segtree[at]; }
+		int mid = (at_left + at_right) / 2;
+		T left_res = range_sum(start, end, 2 * at, at_left, mid);
+		T right_res = range_sum(start, end, 2 * at + 1, mid + 1, at_right);
+		return combine(left_res, right_res);
+	}
+
+  public:
+	SumSegmentTree(int len) : len(len) { segtree = vector<T>(len * 4, DEFAULT); };
+
+	SumSegmentTree(const vector<T> &arr) : len(arr.size()) {
+		segtree = vector<T>(len * 4, DEFAULT);
```

Comment:
any reason why we're using a 4n memory segtree instead of a 2n like before?

Reply:
ask justin lol

Reply:
iterative segtree actually has a kind of funny layout - i’ll find a diagram later

Reply:
![image](https://github.com/user-attachments/assets/7a4b7d81-255a-4fb3-85df-c4a9e
above is what an iterative segtree looks like for n=13. as you can see, the layout is not the same as your typical recursive segtree. (this is also why iterative segtree is way more confusing IMO)

Reply:
also, if you want to reduce memory, `4 << __lg(n)` is a valid length for the array

Reply:
yeah, i think we should leave the iterative versions as an afterthought and mostly focus on the recursive versions

Reply:
hm ok